lib/api/core/models/repositories/repository.js

```diff
@@ -1,344 +1,421 @@
 
-      if (response === null) {
-        return null;
-      }
+            return _.assignIn(object, document._source, {_id: document._id});
+          });
+      });
+  }
 
-      object = new this.ObjectConstructor();
+  /**
+   * Search in database corresponding repository according to a query
+   *
+   * @param {object} query
+   * @param {number} from - manage pagination
+   * @param {number} size - manage pagination
+   * @param {string} [collection]
+   * @returns {Promise}
+   */
+  search (query, from, size, collection = this.collection) {
+    if (collection === null) {
+      return Promise.reject(new InternalError('The collection must be specified.'));
+    }
 
-      return _.assignIn(object, JSON.parse(response));
-    })
-    .catch(err => Promise.reject(new InternalError(err)));
-};
+    return this.databaseEngine.search(collection, query, from, size)
+      .then(response => {
+        if (!response.hits || response.hits.length === 0) {
+          return {total: 0, hits: []};
+        }
 
-/**
- * Loads an object from Cache or from the Database if not available in Cache.
- * Returns a promise that resolves either to the
- * retrieved object of null in case it is not found.
- *
- * If the object is not found in Cache and found in the Database,
- * it will be written to cache also.
- *
- * The opts object currently accepts one optional parameter: key, which forces
- * the cache key to fetch.
- * In case the key is not provided, it defauls to <collection>/id, i.e.: _kuzzle/users/12
- *
- * @param {string} id The id of the object to get
- * @param {object} [opts] Optional options.
- * @returns {Promise}
- */
-Repository.prototype.load = function repositoryLoad (id, opts) {
-  if (!this.cacheEngine) {
-    return this.loadOneFromDatabase(id);
+        return {
+          total: response.total,
+          hits: response.hits.map((document) => {
+            return _.assignIn({}, document._source, {_id: document._id});
+          })
+        };
+      });
   }
 
-  return this.loadFromCache(id, opts)
-    .then(object => {
-      if (object === null) {
-        if (this.databaseEngine === null) {
+  /**
+   * Loads an object from Cache. Returns a promise that resolves either to the
+   * retrieved object of null in case it is not found.
+   *
+   * The opts object currently accepts one optional parameter: key, which forces
+   * the cache key to fetch.
+   * In case the key is not provided, it defaults to repos/<index>/<collection>/<id>, i.e.: repos/%kuzzle/users/12
+   *
+   * @param {string} id - The id of the object to get
+   * @param {object} [options] - Optional options.
+   * @param {string} [collection]
+   * @returns {Promise}
+   */
+  loadFromCache (id, options = {}, collection = this.collection) {
+    const key = options.key || this.getCacheKey(id, collection);
+
+    if (collection === null) {
+      return Promise.reject(new InternalError('The collection must be specified.'));
```

Comment:
It can not happen due to a bad request, it might come from an bad plugin implementation, will use this one instead.